r/src/bpcells-cpp/matrixIterators/H5DenseMatrixWriter.h

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022 BPCells contributors
+// 
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#pragma once
+
+#include "../arrayIO/array_interfaces.h"
+#include "../arrayIO/bp128.h"
+#include "../arrayIO/hdf5.h"
+#include "../utils/filesystem_compat.h"
+#include "MatrixIterator.h"
+#include "OrderRows.h"
+
+namespace BPCells {
+
+template <typename T> class H5DenseMatrixWriter : public MatrixWriter<T> {
+  private:
+    HighFive::DataSet dataset;
+    HighFive::DataType datatype = HighFive::create_datatype<T>();
+    uint64_t max_capacity;
+    bool row_major;
+
+    static HighFive::DataSet createH5Matrix(
+        HighFive::File file,
+        std::string dataset_path,
+        uint64_t nrow,  // nrow of input matrix
+        uint64_t ncol,  // ncol of input matrix
+        bool row_major,
+        uint64_t chunk_size,
+        uint32_t gzip_level
+    ) {
+      HighFive::SilenceHDF5 s;
+      // Create a dataspace with initial shape and max shape
+      uint64_t nrow_h5 = nrow;
+      uint64_t ncol_h5 = ncol;
+      if (row_major) {
+          nrow_h5 = ncol;
+          ncol_h5 = nrow;
+      }
+      HighFive::DataSpace dataspace({nrow_h5, ncol_h5}, {HighFive::DataSpace::UNLIMITED, HighFive::DataSpace::UNLIMITED});
+
+      // Use chunking
+      HighFive::DataSetCreateProps props;
+      props.add(HighFive::Chunking(std::vector<hsize_t>{std::min(chunk_size, nrow_h5), std::min(chunk_size, ncol_h5)}));
+      if (gzip_level > 0) {
+          props.add(HighFive::Shuffle());
+          props.add(HighFive::Deflate(gzip_level));      
+      }
+
+      if (file.exist(dataset_path)) {
+          file.unlink(dataset_path);
+      }
```

Comment:
I don't BPCells to ever default to overwriting existing data. Either we should just allow HighFive to throw a C++ error like `write_matrix_anndata_hdf5` currently does, or we can add an `overwrite = FALSE` parameter in R like `write_matrix_dir` does (and only do unlink if `overwrite == TRUE`)